ecs-agent/wsclient/client.go

```diff
@@ -0,0 +1,578 @@
+// to be interface{} to properly capture that
+type RequestHandler interface{}
+
+// RequestResponder wraps the RequestHandler interface with a Respond()
+// method that can be used to Respond to requests read and processed via
+// the RequestHandler interface for a particular message type.
+//
+// Example:
+//
+//	type payloadMessageDispatcher struct {
+//	    respond  func(interface{}) error
+//	    dispatcher actor.Dispatcher
+//	}
+//	func(d *payloadmessagedispatcher) RegisterResponder(respond func(interface{}) error) error {
+//	    d.respond = respond
+//	    return nil
+//	}
+//	func(d *payloadmessagedispatcher) HandlerFunc() RequestHandler {
+//	    return func(payload *ecsacs.PayloadMessage) {
+//	        message := &actor.DispatcherMessage{
+//	            Payload: payload,
+//	            AckFunc: func() error {
+//	                return d.respond()
+//	            },
+//	            ...
+//	        }
+//	        d.dispatcher.Send(message)
+//	    }
+//	}
+type RequestResponder interface {
+	// Name returns the name of the responder. This is used mostly for logging.
+	Name() string
+	// RegisterResponder registers a function that can be invoked in response
+	// to receiving and processing a websocket request message.
+	RegisterResponder(RespondFunc)
+	// HandlerFunc returns the RequestHandler callback for a particular
+	// websocket request message type.
+	HandlerFunc() RequestHandler
+}
+
+// RespondFunc specifies a function callback that can be used by the
+// RequestResonder to respond to requests.
+type RespondFunc func(interface{}) error
+
+// ClientServer is a combined client and server for the backend websocket connection
+type ClientServer interface {
+	AddRequestHandler(RequestHandler)
+	// SetAnyRequestHandler takes a function with the signature 'func(i
+	// interface{})' and calls it with every message the server passes down.
+	// Only a single 'AnyRequestHandler' will be active at a given time for a
+	// ClientServer
+	SetAnyRequestHandler(RequestHandler)
+	MakeRequest(input interface{}) error
+	WriteMessage(input []byte) error
+	WriteCloseMessage() error
+	Connect() error
+	IsConnected() bool
+	SetConnection(conn wsconn.WebsocketConn)
+	Disconnect(...interface{}) error
+	Serve(ctx context.Context) error
+	SetReadDeadline(t time.Time) error
+	io.Closer
+}
+
+// AgentConfig is a subset of agent's config. Since this requires agent specific information,
+// making this as a struct to be passed from the individual agents while using the wsclient shared lib.
+// The agent where the field is not applicable can be left empty
+type WSClientMinAgentConfig struct {
+	AWSRegion          string
+	AcceptInsecureCert bool
+	DockerEndpoint     string
+	IsDocker           bool
+}
+
+// ClientServerImpl wraps commonly used methods defined in ClientServer interface.
+type ClientServerImpl struct {
+	// Cfg is the subset of user-specified runtime configuration
+	Cfg *WSClientMinAgentConfig
```

Comment:
Does Cfg need to be exported?

Reply:
Yes, it would need to be populated by the ACS/TACS libraries. So it needs to be exported.